kaggle_environments/envs/lux_ai_2022/luxai2022/map/board.py

```diff
@@ -0,0 +1,124 @@
+
+        # remove bottom once generator is fully ready
+        self.map.rubble = self.map.rubble.astype(int)
+        self.map.ore = self.map.ore.astype(int)
+        self.map.ice = self.map.ice.astype(int)
+
+        self.lichen = np.zeros((self.height, self.width), dtype=int)
+        # ownership of lichen by factory id, a simple mask
+        # -1 = no ownership
+        self.lichen_strains = -np.ones((self.height, self.width), dtype=int)
+        # self.units_map: np.ndarray = np.zeros((self.height, self.width))
+        self.units_map: Dict[str, List[Unit]] = defaultdict(list)
+        # TODO consider unit occupancy map, may speed up computation by storing more info in a form indexable in numpy arrays
+
+        # maps center of factory to the factory
+        self.factory_map: Dict[str, 'Factory'] = dict()
+        # != -1 if a factory tile is on the location. Equals factory number id
+        self.factory_occupancy_map = -np.ones((self.height, self.width), dtype=int)
+
+
+        # Valid spawn locations
+        player_0_spawn_info = self.get_valid_spawns(0)
+        player_1_spawn_info = self.get_valid_spawns(1)
+        self.spawns = {"player_0": player_0_spawn_info[0],
+                       "player_1": player_1_spawn_info[0]}
+        self.spawn_masks = {"player_0": player_0_spawn_info[1],
+                       "player_1": player_1_spawn_info[1]}
+    def pos_hash(self, pos: Position):
+        return f"{pos.x},{pos.y}"
+    def get_units_at(self, pos: Position):
+        pos_hash = self.pos_hash(pos)
+        if pos_hash in self.units_map:
+            return self.units_map[pos_hash]
+        return None
+    def get_factory_at(self, pos: Position):
+        pos_hash = self.pos_hash(pos)
+        if pos_hash in self.factory_map:
+            return self.factory_map[pos_hash]
+        return None
+    def get_valid_spawns(self, team_id):
+        xx, yy = np.mgrid[:self.width, :self.height]
+        if self.map.symmetry == "vertical":
+            if team_id == 0:
+                spawns_mask = xx < (self.width - 1) / 2
+            else:
+                spawns_mask = xx > (self.width - 1) / 2
+        if self.map.symmetry == "horizontal":
+            if team_id == 0:
+                spawns_mask = yy < (self.height - 1) / 2
+
+            else:
+                spawns_mask = yy > (self.height - 1) / 2
+
+        if self.map.symmetry == "rotational":
+            if team_id == 0:
+                spawns_mask = xx < (self.width - 1) / 2
+
+            else:
+                spawns_mask = xx > (self.width - 1) / 2
+
+        if self.map.symmetry == "/":
+            if team_id == 0:
+                spawns_mask = xx - yy < 0
+
+            else:
+                spawns_mask = xx - yy > 0
+
+        if self.map.symmetry == "\\":
+            if team_id == 0:
+                spawns_mask = xx + yy < (self.width + self.height) / 2 - 1
+
+            else:
+                spawns_mask = xx + yy > (self.width + self.height) / 2 - 1
```

Comment:
nice! hard to validate this without checking, but did you run through these to make sure they are correct?

Reply:
both symmetries work fine!

Reply:
I've actually temporarily disabled the diagonal symmetries for now. Since it's just a map generation thing we can usually add it later.